File: tests/config/configParser.test.ts
import { describe, it, expect } from "vitest";
import {
  defaultAppLocalConfig,
  defaultParentPrivateConfig,
  defaultGroupSharedConfig,
  defaultSyncConfig,
  defaultUiPreferenceConfig,
  parseConfig,
  encodeConfig,
  type AppLocalConfig,
} from "@/domain/config";
import { AppLanguage, LandingScreen, ThemeMode } from "@/domain/enums";

describe("configParser round-trips", () => {
  it("AppLocalConfig round-trips", () => {
    const original: AppLocalConfig = {
      ...defaultAppLocalConfig,
      themeMode: ThemeMode.DARK,
      language: AppLanguage.HEBREW,
      defaultLandingScreen: LandingScreen.RIDES,
      syncOnAppOpen: false,
      backgroundSyncEnabled: false,
      backgroundSyncIntervalMinutes: 60,
      showCompletedRidesByDefault: true,
      compactCardMode: true,
      vibrateOnReminder: false,
      soundOnReminder: false,
      notificationLeadTimeMinutesDefault: 30,
      allowMobileDataSync: false,
      debugLoggingEnabled: true,
      lastSelectedGroupId: "group-123",
    };
    const restored = parseConfig(encodeConfig(original), defaultAppLocalConfig);
    expect(restored).toEqual(original);
  });

  it("ParentPrivateConfig round-trips with defaults", () => {
    const text = encodeConfig(defaultParentPrivateConfig);
    expect(parseConfig(text, defaultParentPrivateConfig)).toEqual(defaultParentPrivateConfig);
  });

  it("GroupSharedConfig round-trips with custom values", () => {
    const cfg = {
      ...defaultGroupSharedConfig,
      groupName: "Soccer Parents",
      groupDescription: "Test group",
      timezone: "America/New_York",
      enableRideSharing: true,
      allowMultipleVolunteersPerRideLeg: false,
      maxFutureEventMonths: 6,
    };
    expect(parseConfig(encodeConfig(cfg), defaultGroupSharedConfig)).toEqual(cfg);
  });

  it("SyncConfig round-trips", () => {
    expect(parseConfig(encodeConfig(defaultSyncConfig), defaultSyncConfig)).toEqual(defaultSyncConfig);
  });

  it("UiPreferenceConfig round-trips", () => {
    expect(parseConfig(encodeConfig(defaultUiPreferenceConfig), defaultUiPreferenceConfig))
      .toEqual(defaultUiPreferenceConfig);
  });

  it("AppLocalConfig ignores unknown keys for forward-compat", () => {
    const payload = JSON.stringify({ themeMode: ThemeMode.DARK, futureField: "ignored" });
    const restored = parseConfig(payload, defaultAppLocalConfig);
    expect(restored.themeMode).toBe(ThemeMode.DARK);
    // @ts-expect-error — futureField is not part of AppLocalConfig
    expect(restored.futureField).toBeUndefined();
  });
});

File: app/src/test/java/i/drive/kids/config/ConfigParserTest.kt
package i.drive.kids.config

import i.drive.kids.domain.model.AppLanguage
import i.drive.kids.domain.model.LandingScreen
import i.drive.kids.domain.model.ThemeMode
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.junit.Assert.assertEquals
import org.junit.Test

/**
 * Round-trip serialization tests for all four config scopes.
 * Ensures no field is lost when a config survives write→read cycle (which is how
 * it's persisted to Drive JSON / Sheet / DataStore).
 */
class ConfigParserTest {

    private val json = Json {
        ignoreUnknownKeys = true
        prettyPrint = false
        encodeDefaults = true
    }

    @Test
    fun `AppLocalConfig round-trips`() {
        val original = AppLocalConfig(
            themeMode = ThemeMode.DARK,
            language = AppLanguage.HEBREW,
            defaultLandingScreen = LandingScreen.RIDES,
            syncOnAppOpen = false,
            backgroundSyncEnabled = false,
            backgroundSyncIntervalMinutes = 60,
            showCompletedRidesByDefault = true,
            compactCardMode = true,
            vibrateOnReminder = false,
            soundOnReminder = false,
            notificationLeadTimeMinutesDefault = 30,
            allowMobileDataSync = false,
            debugLoggingEnabled = true,
            lastSelectedGroupId = "group-123"
        )
        val text = json.encodeToString(original)
        val restored = json.decodeFromString<AppLocalConfig>(text)
        assertEquals(original, restored)
    }

    @Test
    fun `ParentPrivateConfig round-trips with defaults`() {
        val original = ParentPrivateConfig()
        val text = json.encodeToString(original)
        val restored = json.decodeFromString<ParentPrivateConfig>(text)
        assertEquals(original, restored)
    }

    @Test
    fun `GroupSharedConfig round-trips with custom values`() {
        val original = GroupSharedConfig(
            groupName = "Soccer Parents",
            groupDescription = "Test group",
            timezone = "America/New_York",
            enableRideSharing = true,
            allowMultipleVolunteersPerRideLeg = false,
            maxFutureEventMonths = 6
        )
        val text = json.encodeToString(original)
        val restored = json.decodeFromString<GroupSharedConfig>(text)
        assertEquals(original, restored)
    }

    @Test
    fun `SyncConfig round-trips`() {
        val original = SyncConfig()
        val text = json.encodeToString(original)
        val restored = json.decodeFromString<SyncConfig>(text)
        assertEquals(original, restored)
    }

    @Test
    fun `UiPreferenceConfig round-trips`() {
        val original = UiPreferenceConfig()
        val text = json.encodeToString(original)
        val restored = json.decodeFromString<UiPreferenceConfig>(text)
        assertEquals(original, restored)
    }

    @Test
    fun `AppLocalConfig ignores unknown keys for forward-compat`() {
        val payload = """{"themeMode":"DARK","futureField":"ignored"}"""
        val restored = json.decodeFromString<AppLocalConfig>(payload)
        assertEquals(ThemeMode.DARK, restored.themeMode)
    }
}
